lib/compress/zstd_compress.c

```diff
@@ -2672,6 +2734,593 @@ static void ZSTD_confirmRepcodesAndEntropyTables(ZSTD_CCtx* zc)
+    literalsBytes = ZSTD_countSeqStoreLiteralsBytes(resultSeqStore);
+    resultSeqStore->litStart += literalsBytesPreceding;
+    if (endIdx == (size_t)(originalSeqStore->sequences - originalSeqStore->sequencesStart)) {
+        /* This accounts for possible last literals if the derived chunk reaches the end of the block */
+        resultSeqStore->lit = litEnd;
+    } else {
+        resultSeqStore->lit = resultSeqStore->litStart+literalsBytes;
+    }
+    resultSeqStore->llCode += startIdx;
+    resultSeqStore->mlCode += startIdx;
+    resultSeqStore->ofCode += startIdx;
+}
+
+/* ZSTD_compressSequences_singleBlock():
+ * Compresses a seqStore into a block with a block header, into the buffer dst.
+ * 
+ * Returns the total size of that block (including header) or a ZSTD error code.
+ * 
+ * TODO: Migrate compressBlock_internal and compressSequences_internal to use this as well
+ */
+static size_t ZSTD_compressSequences_singleBlock(ZSTD_CCtx* zc, seqStore_t* seqStore,
+                                                 void* dst, size_t dstCapacity,
+                                                 const void* src, size_t srcSize,
+                                                 U32 lastBlock) {
+    const U32 rleMaxLength = 25;
+    BYTE* op = (BYTE*)dst;
+    const BYTE* ip = (const BYTE*)src;
+    size_t cSize;
+    size_t cSeqsSize = ZSTD_entropyCompressSequences(seqStore,
+                &zc->blockState.prevCBlock->entropy, &zc->blockState.nextCBlock->entropy,
+                &zc->appliedParams,
+                op + ZSTD_blockHeaderSize, dstCapacity - ZSTD_blockHeaderSize,
+                srcSize,
+                zc->entropyWorkspace, ENTROPY_WORKSPACE_SIZE /* statically allocated in resetCCtx */,
+                zc->bmi2);
+
+    if (!zc->isFirstBlock &&
+        cSeqsSize < rleMaxLength &&
+        ZSTD_isRLE((BYTE const*)src, srcSize)) {
+        /* We don't want to emit our first block as a RLE even if it qualifies because
+        * doing so will cause the decoder (cli only) to throw a "should consume all input error."
+        * This is only an issue for zstd <= v1.4.3
+        */
+        cSeqsSize = 1;
+    }
+
+    if (zc->seqCollector.collectSequences) {
+        ZSTD_copyBlockSequences(zc);
+        return 0;
+    }
+
+    if (zc->blockState.prevCBlock->entropy.fse.offcode_repeatMode == FSE_repeat_valid)
+        zc->blockState.prevCBlock->entropy.fse.offcode_repeatMode = FSE_repeat_check;
+
+    if (cSeqsSize == 0) {
+        cSize = ZSTD_noCompressBlock(op, dstCapacity, ip, srcSize, lastBlock);
+        FORWARD_IF_ERROR(cSize, "Nocompress block failed");
+        DEBUGLOG(4, "Writing out nocompress block, size: %zu", cSize);
+    } else if (cSeqsSize == 1) {
+        cSize = ZSTD_rleCompressBlock(op, dstCapacity, *ip, srcSize, lastBlock);
+        FORWARD_IF_ERROR(cSize, "RLE compress block failed");
+        DEBUGLOG(4, "Writing out RLE block, size: %zu", cSize);
+    } else {
```

Comment:
It is potentially a bug to use either of these compression modes.

These modes change the repcode history. So it would only be safe to use them if you can guarantee that repcodes don't reference sequences in this block. A sufficient condition would be to check that the next 3 sequences aren't repcodes (and there are at least 3 remaining sequences). A safer approach would be to not allow them at all.

Reply:
Can you add a test case that catches this, or verify that the oss-fuzz fuzzers can catch this?

I believe you don't see this trivially because you don't split blocks of less than 300 sequences. That makes no-compress blocks or rle blocks less likely, but not impossible.

Reply:
So the fuzzer seems to catch this fairly easily when I reduce this constant down to 20, but when it's at 300, running it for a few hours doesn't catch it (though I guess at some point it might get caught anyways).

For a unit test, I'm having trouble trying to recreate this scenario in a unit test - a block of 300 sequences would need generated to be such that the first 150 sequences must be uncompressible, followed by a sequence that uses repcode history to reference an offset from one of the 150 sequences (corrupted), followed by 150 very compressible sequences (to incentivize block splitting).

How would someone generate a series of uncompressible sequences?

Reply:
Maybe https://en.wikipedia.org/wiki/De_Bruijn_sequence with 4-byte runs of zeros interspersed (for matches)?